src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp

```diff
@@ -0,0 +1,387 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines which types are allowed, whether operations which certain types are
+/// allowed, and other type-specific properties and configuration for
+/// `TensorExpression`s
+///
+/// \details
+/// To add support for a data type, modify the templates in this file and the
+/// arithmetic operator overloads as necessary. Then, add tests as appropriate.
+
+#pragma once
+
+#include <cstddef>
+#include <limits>
+#include <type_traits>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/VectorImpl.hpp"
+
+namespace tenex {
+template <typename DataType>
+struct NumberAsExpression;
+
+namespace detail {
+/// \brief Whether or not `TensorExpression`s supports using a given type as a
+/// numeric term
+///
+/// \details
+/// To make it possible to use a new numeric data type as a term in
+/// `TensorExpression`s, add the type to this alias and adjust other templates
+/// in this file, as necessary.
+///
+/// \tparam X the arithmetic data type
+template <typename X>
+using is_supported_number_datatype =
+    std::bool_constant<std::is_same_v<X, double>>;
+
+/// \brief Whether or not `Tensor`s with the given data type are currently
+/// supported by `TensorExpression`s
+///
+/// \details
+/// To make it possible to use a new data type in a `Tensor` term in
+/// `TensorExpression`s, add the type to this alias and adjust other templates
+/// in this file, as necessary.
+///
+/// \tparam X the `Tensor` data type
+template <typename X>
+using is_supported_tensor_datatype =
+    std::bool_constant<std::is_same_v<X, double> or
+                       std::is_same_v<X, DataVector>>;
+
+/// \brief Whether or not the given type is a `VectorImpl` type
+///
+/// \tparam T the given type
+template <typename T>
+using is_vector = std::bool_constant<std::is_base_of_v<MarkAsVectorImpl, T>>;
+
+/// \brief If the given type is a derived `VectorImpl` type, get the base
+/// `VectorImpl` type, else return the given type
+///
+/// \tparam T the given type
+/// \tparam IsVector whether or not the given type is a `VectorImpl` type
+template <typename T, bool IsVector = is_vector<T>::value>
+struct upcast_if_derived_vector_type;
+
+/// If `T` is not a `VectorImpl`, the `type` is just the input
+template <typename T>
+struct upcast_if_derived_vector_type<T, false> {
+  using type = T;
+};
+/// If `T` is a `VectorImpl`, the `type` is the base `VectorImpl` type
+template <typename T>
+struct upcast_if_derived_vector_type<T, true> {
+  using upcasted_type = typename T::BaseType;
+  // if we have a derived VectorImpl, get base type, else T is a base VectorImpl
+  // and we use that
+  using type =
+      tmpl::conditional_t<std::is_base_of_v<MarkAsVectorImpl, upcasted_type>,
```

Comment:
I think this needs to stay as is because the key part here is that this `type` should be the base `VectorImpl` type. Since something like `is_derived_of_vector_impl_v<VectorImpl<double, DataVector>>` will be `true`, this would incorrectly assign this `type` to be the base blaze type like `blaze::CustomVector<...>`, which is not what I want

Let me know if I am misunderstanding. I basically want this to be able to take a type and if it is a derived `VectorImpl`, I want the base, otherwise do nothing to the type.